ユーザー管理機能

AG0811 opened the pull request:
＃What
素材より、CSSを移動(動作確認をしやすくするため)
deviseを導入
素材より、registrations/new.html.erb移動・修正
素材より、sessions/new.html.erb移動・修正

＃Why
ユーザー管理機能を実装するため

◆必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/8345dab925cb1e4fb45afc306e179916

◆入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/bc36eff110edf4242797da63c15888d2

◆必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/3823273ef1a7b2335460e48aab0a4643

◆入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/9533152feadcb0535ce62fbf5edbbc93

◆ログアウトに成功する動画
https://gyazo.com/ec59fd42f94d208cc62451f9729f727f

◆テストに成功する画像
https://gyazo.com/d4097f2cd6aec1591a899edaaba6c80d

AG0811 committed:
devise導入～ログイン画面表示迄

AG0811 committed:
単体テスト実装

ユーザー管理機能のテストケース作成

maimai0401 commented on app/controllers/application_controller.rb:
```diff
@@ -11,4 +12,8 @@ def basic_auth
     end
   end
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname, :first_name, :last_name, :read_first, :read_last, :birthday])
```
devise_parameter_sanitizerメソッドを一つにまとめましょう。

■理由
重複した内容はリファクタリングされることが好ましいためです。

maimai0401 commented on app/models/user.rb:
```diff
@@ -0,0 +1,20 @@
+
+  # 本人情報確認のバリデーション
+  with_options presence: true do
+    validates :password, presence: true, length: { minimum: 6 },format: { with: /\A(?=.*?[a-z])(?=.*?\d)[a-z\d]+\z/i, message: 'is invalid. Include both letters and numbers' }
```
passwordに不要なバリデーションが付与されていないか確認しましょう。
空や6文字未満についてはデフォルトで制限されているためバリデーションは不要です。

maimai0401 commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,109 @@
+<%= render "shared/second-header"%>
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

maimai0401 commented on app/views/shared/_header.html.erb:
```diff
@@ -15,13 +15,16 @@
 
-      <li><%= link_to 'ログイン', "#", class: "login" %></li>
-      <li><%= link_to '新規登録', "#", class: "sign-up" %></li>
+      <%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

maimai0401 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,127 @@
+      end
+    end
+  end
+end
```
下記テストを追加で実装しましょう。
・全角文字を含むパスワードでは登録できない

AG0811 replied:
レビューありがとうございます。
上記の旨かしこまりました。
確認し修正します。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください。
どうぞよろしくおねがいします。

AG0811 committed:
レビュー反映分

＃What
devise_parameter_sanitizerメソッド内の内容をリファクタリングで完結に修正
passwordに元から設定されている6文字以上というバリデーションを消去
htmlに記述されている装案内コメントを実装済みであるならば消去
単体テストコードにPasswordに全角が含まれる場合を追加

＃Why
ユーザー管理機能を実装するため
レビュー内容を反映させるため

AG0811 committed:
再修正

Key名修正

AG0811 commented:
＃What
devise_parameter_sanitizerメソッド内の内容をリファクタリングで完結に修正
passwordに元から設定されている6文字以上というバリデーションを消去
htmlに記述されている装案内コメントを実装済みであるならば消去
単体テストコードにPasswordに全角が含まれる場合を追加

＃Why
ユーザー管理機能を実装するため
レビュー内容を反映させるため

※重複内容のリファクタリングが含まれているため再提出します
◆【再】必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/c3ed9badda9e09ea87784e916e0f84e8

◆【再】入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/2cd4b7d79251b72d62c1cd74c2cd1f85

◆【再】必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/1e01114d32b19bd1b21a476518af7339

◆【再】入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/90f245d1a501f68cece1a1e99bd581da

◆【再】ログアウトに成功する動画
https://gyazo.com/b9be072657368b77e8909862a0767b6f

◆【再】テストに成功する画像
https://gyazo.com/8bb71b425db94b6887d34c9247c77968

wad-sho commented on app/controllers/application_controller.rb:
```diff
@@ -11,4 +12,13 @@ def basic_auth
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: user_keys)
+    devise_parameter_sanitizer.permit(:account_update, keys: user_keys)
+  end
```
devise_parameter_sanitizerメソッドを一つにまとめましょう。
以下のコードは今回のフリマアプリでは実装不要と考えられます。

`
devise_parameter_sanitizer.permit(:account_update, keys: user_keys)`

wad-sho commented on app/models/user.rb:
```diff
@@ -0,0 +1,20 @@
+
+  # 本人情報確認のバリデーション
+  with_options presence: true do
+    validates :password, presence: true,format: { with: /\A(?=.*?[a-z])(?=.*?\d)[a-z\d]+\z/i, message: 'is invalid. Include both letters and numbers' }
```
validates :password, presence: true,format: { with: /\A(?=.*?[a-z])(?=.*?\d)[a-z\d]+\z/i, message: 'is invalid. Include both letters and numbers' }

パスワードのバリデーションにpresence: trueが不要となります。

↑に加え、パスワードのバリデーションはwith_options presence: true do〜endの外で設定する必要がございます。

wad-sho reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いです。
修正の後、再度レビューのご依頼お願い致します！

AG0811 committed:
再々修正

＃What
不要なdevise_parameter_sanitizerメソッドを1つ消去
バリテーションからpasswordからpresence: trueを消去

＃Why
ユーザー管理機能を実装するため
再レビューの内容を反映させるため

AG0811 commented:
＃What
不要なdevise_parameter_sanitizerメソッドを1つ消去
バリテーションからpasswordからpresence: trueを消去

＃Why
ユーザー管理機能を実装するため
再レビューの内容を反映させるため

※パリテーションの修正が含まれているため再提出します
◆【再々】必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/5579453375ec37834011a6bf94115afa

◆【再々】入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/c6d820139ae1dee54f06df20ca067fa6

◆【再々】必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/7f6760688b83cf081e2b2e4d6099987e

◆【再々】入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/b062d632a03696bab3305fb2a9ec490d

◆【再々】ログアウトに成功する動画
https://gyazo.com/afe3ea5ab14df8dbdaee88fb111c3167

◆【再々】テストに成功する画像
https://gyazo.com/a605ed92c92dcda0919cded011c35caa

maimai0401 commented:
レビュー依頼ありがとうございます。

こちら`LGTM`とさせていただきます！実装お疲れさまでした。

【注意】
```ruby
①（次のブランチを切る前に）プルリクエストのMergeとmainブランチへのPullを忘れないようにしましょう。
②次のフェーズに進む際は必ず新しいブランチを作ってから作業しましょう。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
引き続き頑張りましょう！応援しております！

AG0811 commented:
ご確認ありがとうございます。
LGTMを確認いたしました。
此方をマージし、イシューをクローズいたします。
